### DISCUSSION TITLE
Bucket with folder for each user

### DISCUSSION THREAD

**Discussion by sneakysquidd** [supabase/supabase#27278] [Questions] [answered]
I have a bucket and I need to allow users to upload to a folder within the bucket that is specific to that user. However, my current implementation is resulting in the following error 
![image](https://github.com/supabase/supabase/assets/29434916/fec3d606-f2b7-45e8-844e-72abc380dd24)
This is the javascript I am using to upload the images. userID is set to the supabase auth.id
![image](https://github.com/supabase/supabase/assets/29434916/c350f1b9-78b5-4ebd-8074-83bb68d10235)
and I have this policy enabled for all operations currently  
![image](https://github.com/supabase/supabase/assets/29434916/15738186-e907-4bef-90c7-c3ac6c3fb9b1)
I am not sure what is causing this issue.

**Comment by GaryAustin1** [comment]
If you make the RLS just the bucketName does it upload correctly to the folder you want?
Also not clear if your policy is based on ALL or you have that policy for each operation.  That policy will not work for insert by itself.  It has to be WITH CHECK.
If so then you probably don't have a user session in the client at the point you make the call.

**Reply by sneakysquidd** [reply]
Yes if i just have the bucketName it uploads correctly. This is being run on my server so there is not a user session open. What would be the best way to run operations as a user from a server?

**Reply by GaryAustin1** [reply, marked as answer]
Just run as service_role with a supabase-js client (ON A SERVER ONLY) if you know the userID id valid.  That will bypass the RLS.

**Reply by sneakysquidd** [reply]
ok thank you. I plan on only allowing admins to be able to read the files and users can just write to these folders. Could I just change the policy to only allow admins to access them since the service_role will just bypass them anyways?

**Reply by GaryAustin1** [reply]
If you are doing all operations from the server then having no policy is fine.  That would allow no reading (unless a public bucket) and no other operations for the users.  Only a service_role client to do everything.  Otherwise if you want users to upload to only their folder from a browser then you have to have a policy like you have for insert at least (and use WITH CHECK).

**Reply by sneakysquidd** [reply]
So if I want admins to be able to view the images I would just need a select policy that allows that?

**Reply by GaryAustin1** [reply]
I don't know what an admin is to you.   If  you are accessing from a service_role client then RLS is meaningless.  So if admins are using that then no need.  Otherwise if you are using a client with a user session, then yes your RLS needs to check whatever makes them an admin.  Role table, custom claim, etc.
